### ISSUE TITLE
[css-selectors-4] Clarify :lang() behavior when the language range is not a well-formed BCP 47 code

### ISSUE THREAD

**Issue by jfkthame** [w3c/csswg-drafts#8720]
According to https://www.w3.org/TR/selectors-4/#the-lang-pseudo,

> An element’s [content language](https://www.w3.org/TR/css-text-3/#content-language) matches a [language range](https://www.w3.org/TR/selectors-4/#language-range) if, when represented in BCP 47 syntax [[BCP47]](https://www.w3.org/TR/selectors-4/#biblio-bcp47), it matches that language range in an extended filtering operation per [[RFC4647]](https://www.w3.org/TR/selectors-4/#biblio-rfc4647) Matching of Language Tags (section 3.3.2).

The text also goes on to mention that

> The language range does not need to be a *valid language code* to perform this comparison.

[my emphasis] which implies, as I understand it, that something like `:lang("qq")` will match content tagged with `lang="qq"` even though `qq` is not a *valid* language tag (as listed in the IANA registry).

However, the Selectors spec does not specifically address how *ill-formed* (not merely *invalid*) tags should be handled.

According to the language tag syntax given in https://www.rfc-editor.org/rfc/rfc5646#section-2.1, a tag like `åå` (containing non-ASCII characters) would be ill-formed ("the language tags described in this document are sequences of characters from the US-ASCII [[ISO646](https://www.rfc-editor.org/rfc/rfc5646#ref-ISO646)] repertoire"), as would a tag like `en---` (the various subtags following the primary language subtag are *optional*, but the grammar does not allow for them to be *empty*; if they're not present, the corresponding hyphen delimiters should also be omitted).

So how does `:lang()` matching work in the presence of ill-formed codes? It seems to me that a literal reading of the spec requires that such codes *never* match, because its definition of "matches" depends on "when represented in BCP 47 syntax", and such ill-formed codes cannot be represented in BCP 47 at all; they conflict with its basic grammar.

A possible alternative interpretation might be that the handling of ill-formed codes is simply *undefined* (because the spec only addresses what it means to "match" for codes "represented in BCP 47 syntax".

I'm not aware of any compelling use case for ill-formed language codes. So in the interests of clarity and interoperability I would like to ask the WG to confirm (and explicitly note in the spec) that `:lang()` matching is based strictly on BCP 47 and RFC4647, and as such, *ill-formed codes never match*.

(Note that the current implementation in WebKit *does* allow ill-formed tags to match. Thus if content is tagged with `lang="SomeRandomCode-Latn-US"`, which is ill-formed because the primary language subtag is too long, it is nevertheless matched by `:lang(SomeRandomCode)`, `:lang("*-US")`, etc. I think this should be considered a bug in the implementation.)

**Comment by tabatkins** [comment]
Right, this is undefined in the spec right now. If the `:lang()` argument can't be interpreted in BCP-47 syntax, it should definitely just fail to match straight away.

**Comment by zcorpan** [comment]
@tabatkins what if the `:lang()` argument is well-formed language range but the language tag for an element isn't BCT-47 syntax? e.g. `lang="SomeRandomCode-Latn-US"`

**Comment by jfkthame** [comment]
> lang="SomeRandomCode-Latn-US"

That tag cannot be parsed according to BCP47 syntax, and therefore *we have no basis to assume anything about any part of it*, such as the meaning of the `Latn` or `US` substrings in it, or even that the hyphens delimit "subtags" at all.

Therefore, IMO it should not match any `:lang()` pseudo, not even `:lang("*")`.

**Comment by Crissov** [comment]
Twenty years ago, I briefly argued that implementations should be free to use external knowledge about the data format in order map random language information to BCP-47. 

<https://lists.w3.org/Archives/Public/www-style/2003Oct/0234.html>

**Comment by svgeesus** [comment]
Poking this sleeping issue in the context of

 - https://github.com/w3c/csswg-drafts/issues/13469

@tabatkins suggested:

> Right, this is undefined in the spec right now. If the `:lang()` argument can't be interpreted in BCP-47 syntax, it should definitely just fail to match straight away.

Proposed async resolution: Selectors 4 should state that  any `:lang()` argument that cannot be parsed with the BCP-47 syntax causes the selector to not match

**Comment by astearns** [comment]
The CSSWG will automatically accept this resolution one week from now if no objections are raised here. Anyone can add an emoji to this comment to express support. If you do not support this resolution, please add a new comment.

Proposed Resolution: Any `:lang()` argument that cannot be parsed with the BCP-47 syntax causes the selector to not match

**Comment by fantasai** [comment]
@astearns Can you clarify whether the resolution allows `:lang(qq)` to match `lang=qq`?

**Comment by fantasai** [comment]
(I'm generally supportive of whatever @jfkthame proposes, but I think the proposed resolution is unclear what exactly it's proposing.)

**Comment by astearns** [comment]
I believe it would, since while it’s not valid `qq` can still be parsed successfully. @jfkthame or @svgeesus might know better if the proposed resolution needs to be revised to account for this.

**Comment by Crissov** [comment]
`qq` is not invalid [*syntactically*](https://datatracker.ietf.org/doc/html/rfc5646#section-2.1), so `:lang(qq)` should still match `lang=qq` like any other *unassigned* ISO 639 code would. 

<details><summary>RFC 5646 ABNF grammar</summary>

~~~~ abnf
 Language-Tag  = langtag             ; normal language tags
               / privateuse          ; private use tag
               / grandfathered       ; grandfathered tags

 langtag       = language
                 ["-" script]
                 ["-" region]
                 *("-" variant)
                 *("-" extension)
                 ["-" privateuse]

 language      = 2*3ALPHA            ; shortest ISO 639 code
                 ["-" extlang]       ; sometimes followed by
                                     ; extended language subtags
               / 4ALPHA              ; or reserved for future use
               / 5*8ALPHA            ; or registered language subtag

 extlang       = 3ALPHA              ; selected ISO 639 codes
                 *2("-" 3ALPHA)      ; permanently reserved

 script        = 4ALPHA              ; ISO 15924 code

 region        = 2ALPHA              ; ISO 3166-1 code
               / 3DIGIT              ; UN M.49 code

 variant       = 5*8alphanum         ; registered variants
               / (DIGIT 3alphanum)

 extension     = singleton 1*("-" (2*8alphanum))

                                     ; Single alphanumerics
                                     ; "x" reserved for private use
 singleton     = DIGIT               ; 0 - 9
               / %x41-57             ; A - W
               / %x59-5A             ; Y - Z
               / %x61-77             ; a - w
               / %x79-7A             ; y - z

 privateuse    = "x" 1*("-" (1*8alphanum))

 grandfathered = irregular           ; non-redundant tags registered
               / regular             ; during the RFC 3066 era

 irregular     = "en-GB-oed"         ; irregular tags do not match
               / "i-ami"             ; the 'langtag' production and
               / "i-bnn"             ; would not otherwise be
               / "i-default"         ; considered 'well-formed'
               / "i-enochian"        ; These tags are all valid,
               / "i-hak"             ; but most are deprecated
               / "i-klingon"         ; in favor of more modern
               / "i-lux"             ; subtags or subtag
               / "i-mingo"           ; combination
               / "i-navajo"
               / "i-pwn"
               / "i-tao"
               / "i-tay"
               / "i-tsu"
               / "sgn-BE-FR"
               / "sgn-BE-NL"
               / "sgn-CH-DE"

 regular       = "art-lojban"        ; these tags match the 'langtag'
               / "cel-gaulish"       ; production, but their subtags
               / "no-bok"            ; are not extended language
               / "no-nyn"            ; or variant subtags: their meaning
               / "zh-guoyu"          ; is defined by their registration
               / "zh-hakka"          ; and all of these are deprecated
               / "zh-min"            ; in favor of a more modern
               / "zh-min-nan"        ; subtag or sequence of subtags
               / "zh-xiang"

 alphanum      = (ALPHA / DIGIT)     ; letters and numbers

~~~~

</details>

PS: The phrase “cannot be parsed with the BCP-47 syntax” from the proposed resolution by @svgeesus (or equivalently “can't be interpreted …” by @tabatkins) seems semantically identical to “ill-formed” in the initial post of this issue by @jfkthame (or to my use of “syntactically invalid” above).

**Comment by jfkthame** [comment]
Yes, that's my understanding. In order to do any matching, we'd be requiring the tag to be _well-formed_ (so that it's clear how the matching algorithm behaves), but it need not necessarily be _valid_.

As noted in RFC5646:

> A tag is considered "well-formed" if it conforms to the ABNF
>   ([Section 2.1](https://www.rfc-editor.org/rfc/rfc5646.html#section-2.1)).  Language tags may be well-formed in terms of syntax
>   but not valid in terms of content.  However, many operations
>   involving language tags work well without knowing anything about the
>   meaning or validity of the subtags.

The concern in this issue is about _ill-formed_ tags, which IMO cannot meaningfully be matched via RFC 4647. `lang=qq` is _well-formed_, so its matching behavior is well-defined despite not being a _valid_ language tag.

Regarding the proposed resolution:

> Selectors 4 should state that any `:lang()` argument that cannot be parsed with the BCP-47 syntax causes the selector to not match

This only mentions the argument of `:lang()`, and not the content of the `lang` attribute. ISTM that the `lang` attribute must also be well-formed in order for matching to happen. So should we add a second clause to the proposed resolution?

> ... and any `lang` attribute that is not well-formed according to BCP-47 syntax never matches any `:lang()` selector.

Thus, even `:lang("*")` should _not_ match junk such as `lang="something-random"`.

**Comment by astearns** [comment]
OK, resetting the timer for this addition

The CSSWG will automatically accept this resolution one week from now if no objections are raised here. Anyone can add an emoji to this comment to express support. If you do not support this resolution, please add a new comment.

Proposed Resolution: Any `:lang()` argument  or `lang` attribute that cannot be parsed with the BCP-47 syntax causes `:lang()` selectors to not match

**Comment by astearns** [comment]
RESOLVED: Any `:lang()` argument or `lang` attribute that cannot be parsed with the BCP-47 syntax causes `:lang()` selectors to not match

**Comment by svgeesus** [comment]
@jfkthame @aphillips clarifying language, plus two examples, added. Does that resolve this issue?

**Comment by aphillips** [comment]
This won't work.

Language **_tags_** can be well-formed and can be valid. But _extended language ranges_ use a different, much more permissive syntax. See: https://www.rfc-editor.org/rfc/rfc4647.html#section-2.2

Using a language range that does not consist of a well-formed language tag doesn't make it match better, which is what you're trying to get at. 

However, extended language ranges permit the wildcard `*` in any subtag position. Canonically, only the first ("primary language subtag") position matters. Other wildcards normalize away when the range is parsed. But this is complicated because there is no "well-formed-but-language-range" ABNF. Your text won't work and could be considered misleading (does it mean 4647/2.2 or 5646/2.2.9??). It might say something like:

> A _language range_ must be an extended language range according to BCP47.
> Language ranges that are not [well-formed language tags](https://www.w3.org/TR/i18n-glossary/#dfn-well-formed)
> or which would not be a well-formed language tag if an initial wildcard character `*` were replaced with a 
> valid subtag do not match anything.

In practice, it is probably easier to note that ranges that don't match the RFC4647 section 2.2 syntax fail and leave it at that. While ranges like "English" and "1234-banana" are invalid as language tags, they also won't match anything productively. The examples provided illustrate 4647 just as well and don't need to be changed.

**Comment by svgeesus** [comment]
Thanks for the explanation @aphillips  I replaced what was there with your suggested language, keeping the examples unchanged. Please confirm.

**Comment by aphillips** [comment]
@svgeesus LGTM. Thanks.

**Comment by svgeesus** [comment]
WPT were updated already by Chromium WPT export

**Comment by jfkthame** [comment]
> WPT were updated already by Chromium WPT export

I don't think the tests are entirely correct. In particular, https://wpt.live/css/selectors/selectors-4/lang-singleton-subtag-matching.html has tests that expect to match the tags `fr-x` and `zh-x`, but AFAICS according to [BCP47 syntax](https://www.rfc-editor.org/rfc/rfc5646.html#section-2.1) these are ill-formed: the `x` singleton that indicates private-use is _only_ allowed if followed by a further subtag.

**Comment by svgeesus** [comment]
Re-opening until all the affected WPT are correct

**Comment by svgeesus** [comment]
@felipeerias I believe you were updating those tests?

**Comment by felipeerias** [comment]
@svgeesus @aphillips Yes, I recently updated those tests as part of the ongoing Chromium implementation.

My understanding is that:

- extended language ranges in the `:lang()` selector follow the syntax in [RFC4647 2.2](https://datatracker.ietf.org/doc/html/rfc4647#section-2.2);
- content language values (e.g. the `lang` attribute) follow "BCP 47" syntax, which here means specifically [RFC5646 2.1](https://datatracker.ietf.org/doc/html/rfc5646#section-2.1);
- matching follows the extended filtering algorithm in [RFC4647 3.3.2](https://datatracker.ietf.org/doc/html/rfc4647#section-3.3.2).

Values that are not well-formed according to their respective syntax would then never match, which would resolve this issue.

If that's the intention of the spec, then this sentence in the current text adds a contradictory restriction:

> Language ranges that are not well-formed language tags or which would not be a well-formed language tag if an initial wildcard character "*" were replaced with a valid subtag, do not match anything.

I am not sure that this restriction is needed.

For example, `fr-x` is a valid language range (RFC4647) but an invalid language tag (RFC5646). Following the points above:

- `:lang(fr-x)` would **not** match an element with `lang="fr-x"` (invalid tag)
- `:lang(fr-x)` would match an element with `lang="fr-x-standard"` (valid tag)
- `:lang(fr-x)` would **not** match an element with `lang="fr-x-ninechars"` (invalid tag, subtag too long)

**Comment by aphillips** [comment]
@felipeerias, I tend to agree: garbage into the match algorithm produces the right level of not matching without the need for additional work by the spec or implementers. Your tests look good to me.

(Note: The text you quoted would forbid the range `fr-x` from matching because `fr-x` is not a well-formed tag. This is an error in the text I suggested.)

RFC4647 extended filtering does **not** require that implementations check whether language **_tags_** are well-formed (and certainly not valid). An implementation therefore would tend to make all three of your examples "match", even though the first and third example's language tags are not well-formed. However, CSS has chosen to canonicalize language tags and this implies it would fail non-well-formed tags:

> Both the content language and the language range must be canonicalized and converted to extlang form as per section 4.5 of [[RFC5646]](https://www.w3.org/TR/selectors-4/#biblio-rfc5646) prior to the extended filtering operation.

This requirement effectively rejects not only badly formed ranges (`en=fubar`, `åå`, `fr-ninechars`), but also badly formed tags/ranges (`de-de-latn`, `*-10-oboe`) and probably invalid tags.

Normally, extended filtering does not prevent non-well-formed tags from being matched. Nor does it prevent non-well-formed (as language tags) ranges from matching tags (including both well-formed and non-well-formed ones).

Upon reflection, I think this additional well-formedness/validity requirement text should go away. But the implications of the canonicalization step need to be spelled out. Perhaps:

> Both the content language and the language range must be canonicalized and converted to extlang form as per section 4.5 of [[RFC5646]](https://www.w3.org/TR/selectors-4/#biblio-rfc5646) prior to the extended filtering operation. Language tags that are not valid cannot be matched and language ranges that cannot be canonicalized do not match anything.

**Comment by svgeesus** [comment]
@aphillips 

> (Note: The text you quoted would forbid the range `fr-x` from matching because `fr-x` is not a well-formed tag. This is an error in the text I suggested.)

Please correct that text, in this issue, so I can be sure to correct it in the way you have in mind.

> Upon reflection, I think this additional well-formedness/validity requirement text should go away.

Please quote, and then indicate the deletion, of the text that should go away.

> But the implications of the canonicalization step need to be spelled out.

Happy to add that but would prefer to see all the edits agreed in advance, so the issue can be fully resolved and closed.

**Comment by cdoublev** [comment]
@svgeesus, I think the texts that you think should be removed and corrected are the same text.

I am also making these suggestions as more general clarifications:

  > The `:lang()` pseudo-class represents an element whose [content language](https://drafts.csswg.org/css-text-4/#content-language) matches one of the comma-separated language ranges listed in its argument.
  >
  > Each **language range** in `:lang()` must be a valid CSS `<ident>` or `<string>`. (Thus language ranges containing asterisks, for example, must be either correctly escaped or quoted as strings, e.g. `:lang(\*-Latn)` or `:lang("*-Latn")`.)
  >
  > The element’s content language matches a language range if its content language [the same as what is already present...]. Language tags that are not valid cannot be matched and language ranges that cannot be canonicalized do not match anything.

(1) A **language range** is no longer initially defined as an *extended* content language, which would imply that it is received as well-formed and canonicalized.

(2) The CSS syntax requirements appear before those relating to matching the content language ranges/tag.

(3) The text about how to determine the content language of an HTML element is removed, as it already appears in CSS Text, which is referenced by the first occurrence of "content language".

**Comment by cdoublev** [comment]
> Normally, extended filtering does not prevent non-well-formed tags from being matched.

But canonicalization does in RFC 5646 - Section 4.5 (Canonicalization of Language Tags):

  > A language tag is in 'extlang form' when the tag is well-formed according to the rules in Sections 2.1 and 2.2

This implies that the canonicalization ensures that the language tag/ranges are well-formed. But if anything should be validated as well-formed, I think only the language tag needs to be.

**Comment by svgeesus** [comment]
@cdoublev would you like to make a PR with your suggested corrections, which can then be reviewed by @aphillips ?

**Comment by cdoublev** [comment]
Sure. But there is still a disagreement between @jfkthame ([comment](https://github.com/w3c/csswg-drafts/issues/8720#issuecomment-4297838351)) and @aphillips + @felipeerias ([comment](https://github.com/w3c/csswg-drafts/issues/8720#issuecomment-4313197993)) about whether the tag `fr-x-standard` should match the range `fr-x`.

I think everyone agrees that:

  1. `aa-c-cc-b-bb` must match `aa-b-bb-c-cc` and vice-versa
  2. `art-lojban` must match `jbo` and vice-versa
  3. `ar-aao` must match `aao` and vice-versa

(1) and (2) requires canonicalizing the tag and the range. (3) requires transforming into the extended form, which applies then partially undoes (2).

**Note:** basically, canonicalizing requires (i) sorting extension subtags (`b-bb` and `c-cc`) in alphanumeric order and (ii) replacing some subtags with a "preferred" value, while transforming *a tag* into the extended form requires applying then partially reverting (ii).

As noted in previous comments, the [canonicalization algorithm](https://www.rfc-editor.org/rfc/rfc5646#section-4.5), which is explicitly intended for tags, implies that what is canonicalized should be well-formed, with "well-formed" meaning that it matches the appropriate syntax. If I am not mistaken, the only [mention](https://www.rfc-editor.org/rfc/rfc4647#section-4.1) of a "well-formed range" refers to the syntax of tags:

  > Users SHOULD select language ranges that are well-formed, valid language tags according to RFC4646 (substituting wildcards as appropriate in extended language ranges).

So `fr-x` would not be a well-formed range and could not be matched by anything.

However, I also tend to agree with @aphillips and @felipeerias, because `fr-x-standard` matching `fr-x` follows the general intention of the [matching (filtering) algorithm](https://www.rfc-editor.org/rfc/rfc4647#section-3.3.2). For example, `aa-bbb` and `aa` are well-formed, regardless of which syntax is considered, and `aa-bbb` matches `aa`.

Besides, when looking at [section 3.2. Implementation Considerations (RFC 4647)](https://www.rfc-editor.org/rfc/rfc4647#section-3.2), the first option allows `fr-x-standard` to match **the original range** `fr-x`:

  > Applications, protocols, or specifications that canonicalize ranges MUST either perform matching operations with both the canonical and original (unmodified) form of the range or MUST also canonicalize each tag for the purposes of comparison.

While transforming the tag and range into the extended form allows "maximum matching flexibility", the RFCs do not define how to transform a range into the extended form, which is not even considered further below in this same section:

  > Applications, protocols, or specifications that use basic ranges might sometimes receive extended language ranges instead. An application, protocol, or specification MUST choose to:
  >
  >   a) map extended language ranges to basic ranges using the algorithm below,
  >   b) reject any extended language ranges in the language priority list that are not valid basic language ranges, or
  >   c) treat each extended language range as if it were a basic language range, which will have the same result as ignoring them, since these ranges will not match any valid language tags.
  >
  > An extended language range is mapped to a basic language range as follows: if the first subtag is a `*` then the entire range is treated as `*`, otherwise each wildcard subtag is removed. For example, the extended language range `en-*-US` maps to `en-US` (English, United States).

Again, when following (a) or (b), `fr-x-standard` matches `fr-x`. (I do not understand the aforementioned implications of (c).)

Now I do not know if people want to strictly conform to these implementation considerations.

**Comment by aphillips** [comment]
@cdoublev Thanks for the excellent analysis. This response quick-and-dirty ☹️despite its length . Note that I have been working on an addition to I18N's guidelines based on this conversation. [PR Preview](https://deploy-preview-183--bp-i18n-specdev.netlify.app/#lang-matching)

I think this is one of the first serious applications of extended filtering. When extended filtering was included in RFC4647, it was done slightly speculatively. The LTRU WG at IETF more-or-less invented it during the writing process. A few of us wrote implementations so that we didn't _entirely_ ignore half of "rough consensus and running code". It has never been, to my knowledge, widely adopted.

4647 was written with a premise that implementations should not be required to have a copy of the language subtag registry (LSTR), which means that the implementation could work without understanding what the subtags mean. The spec doesn't preclude doing more processing, such as canonicalization. Obviously, if you canonicalize the tags, you also need to do so for the ranges.

> If I am not mistaken, the only [mention](https://www.rfc-editor.org/rfc/rfc4647#section-4.1) of a "well-formed range" refers to the syntax of tags:

The term "well formed" in BCP47 *always* means "conforming to the ABNF grammar without reference to the LSTR". The grammar of both basic and extended language ranges is close to `obs-language-tag` in 5646 (that is, it is like the language tag ABNF in the 1766/3066 era). Well-formed language ranges can include strings that are not in any way valid or even well-formed (non-obsolete) language tags. There is no definition in the RFC for validity in a range, but I think the text we've been working on here is reasonable: it is a valid range if it would be a valid tag once all wildcards are removed (and initial position wildcard replaced with a neutral subtag such as `und`). That is, `*-Latn` is valid (because `und-Latn` is valid), but `*-US-Latn` is not.

Applying canonicalization and other language tag processing to basic ranges seems straightforward: if the range is "valid" (as above), canonicalize as if it were a language tag. If it is not "valid" but is well-formed, use the existing range. 

The extended filtering algorithm probably needs more work (almost to the point where I'd propose writing an additional RFC) to clarify these things. A challenge is that you cannot write a range like `fr-CH` that *excludes* matching the implicit wildcard, e.g. so that it does **not** match `fr-Brai-CH`. 

---

CLDR has gone beyond 4647 in a number of areas, although their focus is mostly on locale handling (and thus on Lookup-related algorithms). Their specification for `addLikelySubtags` seems like it might be important for what CSS wants to use filtering for. For example, if I write `*-Latn`, I probably want it to match `en-US` (`en` has `Suppress-Script: Latn`)

---

With regard to `fr-x` (and friends), the algorithm in 4647 is carefully written so that singletons block the implicit wildcard from matching across the extension/private-use boundary. This prevents range `de-DE` from matching `de-x-DE`, which is important. The range `fr-x` is obviously not a well-formed language _tag_, but is a well-formed language _range_. It has a useful meaning (any French with a private use sequence), so forbidding it as a range seems like overdoing it? The idea of "valid" range either stops once we see a singleton or depends on the spec for that singleton (which we might not want to require). If you're using these subtags, best know what you're doing 🤷 

Note that the algorithm was written to permit such matching of extension/private-use sequences. The implicit wildcard even turns out to more-or-less work with Unicode's `-u` extension (which uses a kind of k-v pattern). A range like `th-u-nu-Thai` (`nu` is numbering system, so this selects for Thai language with Thai digit shapes) will match a tag such as `th-TH-u-ca-gregori-nu-thai`. You can even do stuff like match `*-u-nu` to select tags that have a specific key in them. But you have to include the extension's singleton to match beyond the singleton in a tag.

In any event, I think perhaps work should be done either in BCP47 or in one of I18N or Unicode/CLDR's documents (such as [LTLI](https://www.w3.org/TR/ltli)) or the aforementioned PR to untangle and document these issues for posterity, based on CSS's learnings.

**Comment by svgeesus** [comment]
Thanks all for the valuable discussion. It seems that much work remains to be done, in multiple places. I therefore propose:

- in Selectors 4, to delete the text

> Language ranges that are not well-formed language tags or which would not be a well-formed language tag if an initial wildcard character "*" were replaced with a valid subtag, do not match anything.

- to re-tag this issue for Selectors 5

**Comment by jfkthame** [comment]
> With regard to `fr-x` (and friends), the algorithm in 4647 is carefully written so that singletons block the implicit wildcard from matching across the extension/private-use boundary. This prevents range `de-DE` from matching `de-x-DE`, which is important. The range `fr-x` is obviously not a well-formed language _tag_, but is a well-formed language _range_. It has a useful meaning (any French with a private use sequence), so forbidding it as a range seems like overdoing it?

Agreed that accepting `fr-x` as a range (and therefore allowing it to match `fr-x-foobar` or whatever) is one possible interpretation here, but it does result in a slightly odd situation. While `:lang(fr-x)` will be a "working" selector that matches elements tagged `lang=fr-x-foobar`, it will _fail_ to match an element tagged `lang=fr-x` (because that tag is ill-formed). This might be quite surprising to authors/users?

An alternative would be to treat `fr-x` as an ill-formed range, just like it is an ill-formed tag. To match "any French with a private use sequence", the appropriate range to use would be `fr-x-*`. This avoids the unintuitive situation where a "valid" range fails to match an apparently-identical tag.

**Comment by aphillips** [comment]
@svgeesus 

Rather than removing that text, perhaps it would be better to move it from a normative to an informative posture while we work through the details. Perhaps:

> Language ranges that are not well-formed language tags or which would not be a well-formed language tag if an initial wildcard character "*" were replaced with a valid subtag, often do not match anything or produce undesirable matching.

WRT:
> to re-tag this issue for Selectors 5

I will raise this with I18N in our call tomorrow (@jsahleen for visibility). This issue is currently tracked rather than "needs resolution": we might wish to raise its priority, even while agreeing to slip it. I think it would be okay to move this to Selectors 5, so long as no one is going to cry back compat about it later.

**Comment by aphillips** [comment]
@jfkthame mentioned:
> An alternative would be to treat fr-x as an ill-formed range, just like it is an ill-formed tag. To match "any French with a private use sequence", the appropriate range to use would be fr-x-*. This avoids the unintuitive situation where a "valid" range fails to match an apparently-identical tag.

I agree with you. However... RFC4647 didn't do that. We might impose additional requirements in Selectors, of course, but that would change the range canonicalization operation (the "non-initial wildcards are ignored and thus usually omitted" bits).

**Comment by svgeesus** [comment]
> I will raise this with I18N in our call tomorrow ([@jsahleen](https://github.com/jsahleen) for visibility). This issue is currently tracked rather than "needs resolution": we might wish to raise its priority, even while agreeing to slip it. 

I agree. Note that [Selectors 5](https://www.w3.org/TR/selectors-5/) is being actively worked on; Selectors 4 is the "should have been Rec already" one. This is not a "someday pile".

> I think it would be okay to move this to Selectors 5, so long as no one is going to cry back compat about it later.

I'm hoping that making that sentence informative will avoid that sort of "Selectors 5 contradicts Selectors 4" issue down the line.

**Comment by Crissov** [comment]
@jfkthame I don’t think it would be surprising to those few authors who have to deal with singletons in `lang` in the first place. If I wanted to select all elements that use any Unicode extensions, for instance, I would probably use `:lang("*-u")` intuitively (or the more verbose `:lang("*-u-*")`) and I would **not** want _incomplete / truncated_ attribute values like `lang="en-u"` to match.

**Comment by aphillips** [comment]
We discussed this in the [I18N call today](https://www.w3.org/2026/05/21-i18n-minutes.html#06b9) and will come back again next week to cover more detail. However, the consensus was that slipping to selectors-5 was fine.
